backend/rates/rates.go

```diff
@@ -20,60 +20,124 @@ import (
-	"github.com/digitalbitbox/bitbox-wallet-app/util/socksproxy"
 	"github.com/sirupsen/logrus"
 )
 
+// If modified, also update geckoCoin map.
 var coins = []string{"BTC", "LTC", "ETH", "USDT", "LINK", "MKR", "ZRX", "DAI", "BAT", "USDC"}
+
+// If modified, also update geckoFiat map.
 var fiats = []string{"USD", "EUR", "CHF", "GBP", "JPY", "KRW", "CNY", "RUB", "CAD", "AUD"}
 
 const interval = time.Minute
 const cryptoCompareURL = "https://min-api.cryptocompare.com/data/pricemulti?fsyms=%s&tsyms=%s"
 
-// RateUpdater implements coin.RateUpdater.
+type exchangeRate struct {
+	timestamp time.Time
+	value     float64
+}
+
+// RateUpdater provides cryptocurrency-to-fiat conversion rates.
 type RateUpdater struct {
 	observable.Implementation
-	last       map[string]map[string]float64
+
+	httpClient *http.Client
 	log        *logrus.Entry
-	socksProxy socksproxy.SocksProxy
+
+	// last contains most recent conversion to fiat, keyed by a coin.
+	last map[string]map[string]float64
+
+	historyMu sync.RWMutex
+	// history contains historical conversion rates in asc order, keyed by a coin.
+	// TODO: store in bolt DB
+	// TODO: support different fiat currencies? it's in USD at the moment
+	history map[string][]exchangeRate
+
+	// CoinGecko is where updater gets the historical conversion rates.
+	// See https://www.coingecko.com/en/api for details.
+	coingeckoURL string
 }
 
 // NewRateUpdater returns a new rates updater.
-func NewRateUpdater(socksProxy socksproxy.SocksProxy) *RateUpdater {
-	ratesUpdater := &RateUpdater{
-		last:       map[string]map[string]float64{},
-		log:        logging.Get().WithGroup("rates"),
-		socksProxy: socksProxy,
+func NewRateUpdater(client *http.Client) *RateUpdater {
+	return &RateUpdater{
+		last:         make(map[string]map[string]float64),
+		history:      make(map[string][]exchangeRate),
+		log:          logging.Get().WithGroup("rates"),
+		httpClient:   client,
+		coingeckoURL: coingeckoAPIV3,
 	}
-	go ratesUpdater.start()
-	return ratesUpdater
 }
 
-// Last returns the last rates for a given coin and fiat or nil if not available.
+// Last returns the most recent conversion rates.
+// The returned map is keyed by a crypto coin with values mapped by fiat rates.
+// RateUpdater assumes the returned value is never modified by the callers.
 func (updater *RateUpdater) Last() map[string]map[string]float64 {
 	return updater.last
 }
 
-func (updater *RateUpdater) update() {
-	client, err := updater.socksProxy.GetHTTPClient()
-	if err != nil {
-		updater.log.Printf("Error getting http client %v\n", err)
-		updater.last = nil
-		return
+// PriceUSD returns a historical exchange rate for the given coin.
+// The returned value may be imprecise if at arg matches no timestamp exactly.
+// In this case, linear interpolation is used as an approximation.
+// If no data is available with the given args, PriceUSD returns 0.
+//nolint:staticcheck
+func (updater *RateUpdater) PriceUSD(coin string, at time.Time) float64 {
```

Comment:
what about `PriceAt(coin, fiat string, at time.Time)`? Can panic for now if fiat is not `USD` until implemented.

But once we can switch between currencies, it is easier to pass the fiat as an argument instead of having dedicated functions per fiat currency.

Reply:
Yeah, I had it as `PriceAt` with coin/fiat before and then figured PriceUSD would've been enough for now but you're right. Might as well do it now.

But I'm not a fan of panics. Added a condition to always return 0 for non-USD fiat. This way all the graphs and backend handler can call PriceAt without issues. It'll simply show a flat graph at 0.